### ISSUE TITLE
v3.0.9 calls /authorize in endless loop

### ISSUE THREAD

**Issue by brvaland** [okta/okta-react#34]
Hi,

I am testing okta-react package but noticed since i updated to v3.0.9 my application calls https://dev-xxx.oktapreview.com/oauth2/default/v1/authorize in an endless loop but if i downgrade to v3.0.8 it all works fine.

Router.tsx

```
import React from 'react';
import { Switch, Router, Route } from 'react-router-dom';
import * as createHistory from 'history';
import { Security, SecureRoute, LoginCallback } from '@okta/okta-react';
import OktaAuth from '@okta/okta-auth-js';
import Home from './components/home/Home';

export const history = createHistory.createBrowserHistory();

const config = {
    issuer: `${process.env.REACT_APP_OKTA_ISSUER}`,
    redirectUri: `${window.location.origin}${process.env.REACT_APP_OKTA_REDIRECTURL}`,
    clientId: `${process.env.REACT_APP_OKTA_CLIENTID}`,
    scopes: ['openid', 'email', 'profile'],
    pkce: OktaAuth.features.isPKCESupported()
}

const AppRouter = () => (
    <Router history={history}>
        <div>
            <Security {...config}>
                    <Switch>
                        <SecureRoute path="/" exact={true} component={Home} />
                        <Route path={process.env.REACT_APP_OKTA_REDIRECTURL} component={LoginCallback} />                    
                    </Switch>
            </Security>
        </div>
    </Router>
);

export default AppRouter;
```

**Comment by tomspeak** [comment]
Upgrading to 3.0.9 broke our production build and triggered the same issue. We have had to downgrade to 3.0.8.

**Comment by aarongranick-okta** [comment]
@brvaland @tomspeak Thank you for reporting this issue. I have been able to reproduce the bug and am working on a fix, which we will test and release as soon as possible. In the meantime, I recommend downgrading to version 3.0.8.

**Comment by aarongranick-okta** [comment]
@brvaland @tomspeak We have released `okta-react` version `3.0.10` which should fix this issue.

**Comment by tomspeak** [comment]
thanks, I will try now.

do you do any kind of manual smoke testing? e2e tests? this was a rough one because the bug does not show itself if you still had an active session.

Would like to hear about what can be done to ensure this does not happen again, trust in okta took a hit after this minor patch release going out.

**Comment by swiftone** [comment]
@tomspeak - we have automated unit and e2e testing that must pass before any change is merged, much less gets released, however in this case the tests that would have caught this also used a custom authentication handler that happened to prevent this error from triggering. 

The fix that @aarongranick-okta added (https://github.com/okta/okta-react/pull/37/files) also adds a new e2e test to test without the custom authentication handler to ensure we don't have a repeat of this problem.

We don't like putting out bad releases any more than you like having them - it's literally our job to NOT do that - and we regret that you had that experience.  We plan to continue to improve our automated testing and hope to avoid any repeats of this sort of problem.  That said, mistakes can and will happen and we appreciate your understanding and patience.

**Comment by tomspeak** [comment]
Thank you @swiftone and @aarongranick-okta!

**Comment by fera2k** [comment]
I experienced this issue with version 3.0.10 also. Rolling back to 3.0.8 for now.

**Comment by swiftone** [comment]
@fera2k - Our testing shows 3.0.10 fixes this, so we'd love to get some details on what you're experiencing.  

- are you using the `onAuthRequired` method?
- if so, is that callback part of the loop you see?

**Comment by fera2k** [comment]
Thanks for the reply @swiftone .

When I try to reach a secure route, it starts the infinity loop calls (which all get cancelled status), to this address, and after some seconds, it finally gets redirected. 
https://dev-3333735.okta.com/oauth2/default/v1/authorize?client_id=....

After I put the email and password, it goes back to my CallBack URL on Localhost, with this message:
AuthSdkError: OAuth flow response state doesn't match request state

I'm not using onAuthRequired method and it is working fine on version 3.0.8.

![image](https://user-images.githubusercontent.com/3650296/99293265-2d103200-2821-11eb-8437-2999dddc2b5e.png)

This is my oktaConfig which is being supplied to <Security> element on App.jsx
```
const CLIENT_ID = process.env.REACT_APP_CLIENT_ID;
const OKTA_DOMAIN = process.env.REACT_APP_OKTA_DOMAIN;
const CALLBACK_PATH = '/login/callback';
const HOST = window.location.host;

const ISSUER = process.env.REACT_APP_ISSUER || `https://${OKTA_DOMAIN}/oauth2/default`;
const OKTA_TESTING_DISABLEHTTPSCHECK = process.env.OKTA_TESTING_DISABLEHTTPSCHECK || false;

const oktaConfig = {
  oidc: {
    clientId: CLIENT_ID,
    issuer: ISSUER,
    redirectUri: `http://${HOST}${CALLBACK_PATH}`,
    scopes: ['openid', 'profile', 'email'],
    pkce: true,
    disableHttpsCheck: OKTA_TESTING_DISABLEHTTPSCHECK,
  },
  resourceServer: {
    messagesUrl: 'http://localhost:8000/api/messages',
  },
};

export default oktaConfig;
```

**Comment by chendatao407786** [comment]
We have exact same issue as @fera2k , have to roll back.

**Comment by swiftone** [comment]
@fera2k, @chendatao407786 - Just an update, we are still trying to reproduce the issue you are seeing.  Will update once we have anything else to report.

**Comment by swiftone** [comment]
@fera2k, @chendatao407786 - Can you confirm that you see this infinite loop behavior with okta-react 3.0.10 even in a browser with no existing cookies/localStorage values (perhaps an incognito window?). We have been unable to recreate this loop so far.  

If that doesn't show any difference, can you share any information about how your routes are configured?

**Comment by fera2k** [comment]
Yes @swiftone ,

I'll try to give you as much as detail I can.
I updated my package.json, to use version **3.0.10**, ran `npm install` and double checked the node_modules okta folder. Three dependencies were installed with following versions: 
- okta-auth-js@^3.2.6
- configuration-validation@^0.4.1
- okta-react@^3.0.10

Then I started the project, using npm start and navigate to the root (public route) and double checked that I don't have any LocalStorage or SessionStorage neither cookie before. It's was all empty.

Then I navigated to a secure route. And it is where the infinity loops happens and after hundreds of message, it finally redirects to Okta Login as you can see in the the screenshot , the same as before:
![image](https://user-images.githubusercontent.com/3650296/99460575-daad3f00-290e-11eb-8659-4e810b51d769.png)

In the second time, as my "session" is still active on Okta server, I just got redirected to my callback url, without showing the Okta login page. And I also got this message:
`OAuthError: The response type is not supported by the authorization server. Configured response types: [code].`

Here are the relevant parte of my code:
```
// App.jsx
import React, { useReducer } from 'react';
import { Route, Switch } from 'react-router';
import { Security, LoginCallback } from '@okta/okta-react';
/// ...
import oktaConfig from './config/oktaConfig';

const App = () => {
  const [sessionStore, sessionDispatch] = useReducer(sessionReducer, initialSessionModel);
  const sessionContextValue = { sessionStore, sessionDispatch };

  return (
    <Security {...oktaConfig.oidc}>
      <Switch>
        <Route exact path="/">
          <LayoutPublic>
            <LandingPublic />
          </LayoutPublic>
        </Route>
        <Route path="/login/callback" component={LoginCallback} />
        <SessionContext.Provider value={sessionContextValue}>
          <SecureApp />
        </SessionContext.Provider>
      </Switch>
    </Security>
  );
};
export default App;
```
As you can see, my callback route is public. Besides the public landing page with a header and footer , all the rest are behind secure routes, with I put together in a "SecureApp" file. In this file, I have some code that tries
```
// SecureApp.jsx
import React, { useEffect, useContext } from 'react';
import { Switch } from 'react-router';
import { SecureRoute, useOktaAuth } from '@okta/okta-react';
// ...

const SecureApp = () => {

  const userService = useUserService();

  useEffect(() => {
    const initSession = () => {
      authService.getUser().then((oktaUser) => oktaUser && setOktaInfo(oktaUser));
    };

    const setOktaInfo = (oktaUser) => {
      sessionDispatch({
        type: 'SET_OKTA_INFO',
        data: {
          emailId: oktaUser.preferred_username,
          accessToken: authState.accessToken,
        },
      });
      setUserData(oktaUser.preferred_username);
    };

    const setUserData = async (email) => {
      const response = await userService.getUserData(email);
      if (response && response.ok) {
        const userData = {
           // map data from response.data
        };
        sessionDispatch({ type: 'SET_USER_DATA', data: { userData } });
      }
    };

    if (!authState.isPending && !sessionStore.isUserAuthenticated) {
      initSession();
    }
  }, [
    authService,
    userService,
    authState.accessToken,
    authState.isPending,
    sessionStore.isUserAuthenticated,
    sessionStore.oktaAccessToken,
    sessionDispatch,
  ]);

  if (authState.isPending || sessionStore.isSessionPending) {
    return <div>Loading...</div>;
  }

// ...
  return (
    <Layout>
      <Switch>
        <SecureRoute path="/Home" component={Landing} />
        <SecureRoute path="/MySecureRoute" component={MySecureRouteComponent} />
      </Switch>
    </Layout>
  );
};

export default SecureApp;
```
I hope it helps!

**Comment by swiftone** [comment]
@fera2k - Ah, I think I may see the issue.

TL;DR: 3.0.10 fixes the infinite loop by putting an extra check in `<SecureRoute>` - as you aren't using `<SecureRoute>`, you aren't getting the fix.

Here is the PR for 3.0.10 that addressed this issue for most cases:  https://github.com/okta/okta-react/pull/36

Notice the core issue described: starting the login is updating the state, triggering the loop.  The fix in 3.0.10 puts in a check in SecureRoute to prevent that loop.  (this state change comes from here: https://github.com/okta/okta-react/blob/master/src/AuthService.js#L208).  The upcoming 4.0 release for okta-react shifts isPending to only be used for initial determination on page-load, but until then we have this issue.

I'm talking with the team about options, but if you are looking for an immediate work-around, I would look into breaking up the loop of starting a login via calling .getUser() and then (potentially - I'm looking at your code, still haven't recreated) having .isPending trigger another call - perhaps using a useRef to track if your initSession() is pending, similar to the fix applied to 3.0.10?

**Comment by fera2k** [comment]
I appreciate your help@swiftone. For now I will keep working with version 3.0.8. It's working fine for our app. I will revisit this when realease 4.0 is ready. Keep the good work! 😎

**Comment by shuowu** [comment]
Close the issue, please feel free to reopen it if the issue still exists.
